private/pkg/git/gitbranch/ranger.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitbranch
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/bufbuild/buf/private/pkg/filepathextended"
+	"github.com/bufbuild/buf/private/pkg/git/gitobject"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+)
+
+const defaultRemoteName = "origin"
+
+var baseBranchRefPrefix = []byte("ref: refs/remotes/" + defaultRemoteName + "/")
+
+type rangerOpts struct {
+	baseBranch string
+}
+
+type ranger struct {
+	gitDirPath   string
+	baseBranch   string
+	objectReader gitobject.Reader
+}
+
+func newRanger(
+	gitDirPath string,
+	objectReader gitobject.Reader,
+	opts rangerOpts,
+) (Ranger, error) {
+	gitDirPath = normalpath.Unnormalize(gitDirPath)
+	if err := validateDirPathExists(gitDirPath); err != nil {
+		return nil, err
+	}
+	gitDirPath, err := filepath.Abs(gitDirPath)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: do we need to validate `gitDirPath` is not jumping context?
+	if opts.baseBranch == "" {
+		opts.baseBranch, err = baseBranch(gitDirPath)
+		if err != nil {
+			return nil, fmt.Errorf("automatically determine base branch: %w", err)
+		}
+	}
+	return &ranger{
+		gitDirPath:   gitDirPath,
+		baseBranch:   opts.baseBranch,
+		objectReader: objectReader,
+	}, nil
+}
+
+func (r *ranger) BaseBranch() string {
+	return r.baseBranch
+}
+
+func (r *ranger) Branches(f func(string) error) error {
+	dir := path.Join(r.gitDirPath, "refs", "remotes", defaultRemoteName)
+	return filepathextended.Walk(dir, func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.Name() == "HEAD" || info.IsDir() {
+			return nil
+		}
+		branchName, err := filepath.Rel(dir, path)
+		if err != nil {
+			return err
+		}
+		branchName = normalpath.Normalize(branchName)
+		return f(branchName)
+	})
+}
+
+func (r *ranger) Commits(branch string, f func(gitobject.Commit) error) error {
+	branch = normalpath.Unnormalize(branch)
+	commitBytes, err := os.ReadFile(path.Join(r.gitDirPath, "refs", "remotes", defaultRemoteName, branch))
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return errors.New("branch not found")
```

Comment:
why not report with `fmt.Errorf` what branch was not found? is this a security issue?